File: app/src/main/java/com/example/diffutilrv/view/MainActivity.java
package com.example.diffutilrv.view;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.diffutilrv.R;
import com.example.diffutilrv.databinding.ActivityMainBinding;
import com.example.diffutilrv.veiwmodel.EmployeeSortType;
import com.example.diffutilrv.veiwmodel.MainViewModel;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class MainActivity extends AppCompatActivity {
    @Inject
    public ActivityMainBinding mBinding;
    private MainViewModel mViewModel;
    @Inject
    public EmployeeRecyclerViewAdapter mRecyclerViewAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.mViewModel = new ViewModelProvider(this).get(MainViewModel.class);

        initView();
        initObserver();
    }

    private void initView() {
        this.mBinding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        this.mBinding.recyclerView.setAdapter(mRecyclerViewAdapter);
    }

    private void initObserver() {
        mViewModel.getEmployeeListLiveData().observe(this, employeeList -> mRecyclerViewAdapter.updateEmployeeListItems(employeeList));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.sort_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.sort_by_name:
                mViewModel.updateSortType(EmployeeSortType.NAME);
                return true;
            case R.id.sort_by_role:
                mViewModel.updateSortType(EmployeeSortType.ROLE);
                return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/example/diffutilrv/ui/MainActivity.kt
package com.example.diffutilrv.ui

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.diffutilrv.R
import com.example.diffutilrv.data.EmployeeSortBy
import com.example.diffutilrv.databinding.ActivityMainBinding
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private lateinit var recyclerViewAdapter: EmployeeRecyclerViewAdapter
    private val viewModel by viewModels<MainViewModel>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        recyclerViewAdapter = EmployeeRecyclerViewAdapter(this, viewModel)
        with(binding.recyclerView) {
            layoutManager = LinearLayoutManager(this@MainActivity)
            adapter = recyclerViewAdapter
        }
        observeLiveData()
        if (viewModel.employees.value.orEmpty().isEmpty()) {
            viewModel.loadData()
        }
    }

    private fun observeLiveData() {
        viewModel.employees.observe(this) { recyclerViewAdapter.submitList(it) }
        viewModel.clickRowAction.observe(this) {
            it?.runIfNotConsumed { employee ->
                Toast.makeText(this, getString(R.string.row_clicked_message, employee.name), Toast.LENGTH_SHORT).show()
            }
        }
        viewModel.clickRowButtonAction.observe(this) {
            it?.runIfNotConsumed { employee ->
                Toast.makeText(this, getString(R.string.row_button_clicked_message, employee.name), Toast.LENGTH_SHORT)
                    .show()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.sort_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.sort_by_name -> {
            viewModel.changeSorting(EmployeeSortBy.NAME)
            true
        }
        R.id.sort_by_role -> {
            viewModel.changeSorting(EmployeeSortBy.ROLE)
            true
        }
        else -> super.onOptionsItemSelected(item)
    }
}
